### DISCUSSION TITLE
future of teen_process

### DISCUSSION THREAD

**Discussion by boneskull** [appium/appium#16147] [Ideas]
Should we stop using (and maintaining) `teen_process`?

The widely-used and well-maintained [execa](https://npm.im/execa) seems to replicate what [teen_process](https://npm.im/teen_process) set out to do.

I'm all for reducing maintenance burden, and if we can safely adopt a well-maintained package instead of maintaining our own, that's a win in my book.

However, there are differences in feature-sets between the two.  Here's what I've found:

- `execa` does not allow control of buffer size on a per-stream basis (`maxStdoutBufferSize`, `maxStderrBufferSize`).  We do not seem to use this option in any project under the `appium` org, so I don't believe this is a deal-breaker.
- `execa` has no direct equivalent for `ignoreOutput`, but this could be handled by wrapping the `execa` process' streams the same way we add `data` listeners the spawned process streams, if we need it.  Our projects don't use this option, either.  
- `execa` does not support the `logger` option. This could be implemented via a wrapper, like above.  And again, we don't seem to use this option.
- `execa` does a lot of other nice stuff.  Unclear if there will be a perf hit.  It would be helpful when running e2e tests (more metadata about the process and result), anyway.

While `teen_process` is solid and has absolutely helped many varied Node.js devs, the community may be better served by module consolidation.  If we do decide to end the project, I'd recommend shutting it down entirely instead of attempting to hand it off to a third party for future maintenance (though it probably could not leave the org due to its ownership by the Foundation).

Please lmk your thoughts!

**Comment by jlipps** [comment]
If there's a better maintained alternative I'm open to sunsetting teen process. I'm not necessarily jazzed about updating all of its instances in our codebases, though, which tend to be some of the more complex areas we have. Just to say we'd want to migrate very carefully. Also, the maintenance cost for teen_process has been pretty low relative to how well it has worked for us generally.

**Reply by boneskull** [reply]
In hindsight I think making it a dependency of `@appium/support` and pulling it in that way would have been a good idea.  If we do migrate, all modules should import it (or a wrapper) from `@appium/support`.

**Reply by jlipps** [reply]
maybe so. it was created back in the day when node module authors were encouraged to create lots of small single purpose modules and share them.

**Reply by boneskull** [reply]
No, I don't mean that `teen_process` _should have been a part of `@appium/support`_, it just could be re-exported by `@appium/support`.  Like how we reach in there for filesystem operations

**Reply by jlipps** [reply]
Oh I see. Yes I think you're right.
